File: ts-langchain/src/rag/loader.ts
/**
 * RAG 文档加载器
 * 支持 TXT、Markdown、PDF、DOCX 文件
 */

import * as fs from "fs/promises";
import * as path from "path";
import { parseSessionDocument } from "../services/session-documents.js";

export interface Document {
  id: string;
  content: string;
  metadata: {
    source: string;
    filename: string;
    fileType: string;
    size: number;
    uploadedAt: string;
  };
}

export class DocumentLoader {
  /**
   * 根据文件类型选择加载器
   */
  // 根据文件类型选择合适的加载器读取文档
  static async load(filePath: string, filename: string): Promise<Document> {
    const ext = path.extname(filename).toLowerCase();
    const content = await fs.readFile(filePath, "utf-8");
    const stats = await fs.stat(filePath);

    const doc: Document = {
      id: crypto.randomUUID(),
      content,
      metadata: {
        source: filePath,
        filename,
        fileType: ext,
        size: stats.size,
        uploadedAt: new Date().toISOString(),
      },
    };

    return doc;
  }

  /**
   * 从 Buffer 加载（用于上传场景）
   */
  // 从 Buffer 加载文档（用于上传场景）
  static async loadFromBuffer(
    buffer: Buffer,
    filename: string,
    contentAlreadyParsed = false,
  ): Promise<Document> {
    const ext = path.extname(filename).toLowerCase();
    const content = contentAlreadyParsed
      ? buffer.toString("utf-8")
      : (await parseSessionDocument(buffer, filename, "application/octet-stream"))
          .parts.map((part) => part.content)
          .join("\n\n");

    return {
      id: crypto.randomUUID(),
      content,
      metadata: {
        source: `upload://${filename}`,
        filename,
        fileType: ext,
        size: buffer.length,
        uploadedAt: new Date().toISOString(),
      },
    };
  }
}

File: py-langgraph/src/rag/loader.py
"""
RAG 文档加载器
支持 TXT, Markdown 文件
"""

import os
from dataclasses import dataclass, field


@dataclass
class Document:
    content: str
    metadata: dict = field(default_factory=dict)

    @classmethod
    def from_file(cls, file_path: str) -> "Document":
        """从文件加载文档"""
        filename = os.path.basename(file_path)
        ext = os.path.splitext(filename)[1].lower()

        with open(file_path, "r", encoding="utf-8") as f:
            content = f.read()

        stats = os.stat(file_path)

        return cls(
            content=content,
            metadata={
                "source": file_path,
                "filename": filename,
                "file_type": ext,
                "size": stats.st_size,
                "uploaded_at": __import__("datetime").datetime.now().isoformat(),
            },
        )

    @classmethod
    def from_bytes(cls, content: bytes, filename: str) -> "Document":
        """从字节流加载（用于上传场景）"""
        ext = os.path.splitext(filename)[1].lower()

        return cls(
            content=content.decode("utf-8"),
            metadata={
                "source": f"upload://{filename}",
                "filename": filename,
                "file_type": ext,
                "size": len(content),
                "uploaded_at": __import__("datetime").datetime.now().isoformat(),
            },
        )
